### PR TITLE
Add a ring-buffer reporter to libbeat

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Add a ring buffer reporter that when enabled will store configured
namespaces in a buffer to allow operators to view recent metrics
history. Defaults are to gather the stats namespace every 10s for 10m.
Must be explicitly enabled along with the HTTP endpoints.
The buffer endpoint is intended to be used for diagnostics reporting.

## Why is it important?

We wish to gather recent metrics data with the `elastic-agent diagnostics` command.
The buffer endpoint can be used to collect such data.

Also once this is implemented we may more easily disable the log metrics reporting functionality.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

compile a beat, and run with the following options enabled in the config:

```yaml
http.enabled: true
http.buffer.enabled: true
```

## Related issues

- Relates #29776

Note that this functionality will not be backported, the number of cases metrics could help debug are minimal.

### CODE REVIEW

**Comment by ph** [comment]
This is interesting @michel-laterman let me know if it's ready to look at it.

**Comment by michel-laterman** [comment]
Finally got around to running memory/cpu usage tests:
Running metricbeat 8.1 on ubuntu 18.04 (left) and 20.04 (right).

When http + buffer is enabled on 20.04 we see:
buffered-view
the mb running with the buffer has an rss size of 162-163MB (same as the 18.04 instance without enabling the buffer).

Just enabling the HTTP endpoint on 20.04 we can see the rss size drop to 159-160MB (share average also drops by around 1MB):
http-no-buffer-view

And disabling both the buffer and endpoint leads to an rss size of around 159MB:
no-http-view

CPU usage does not meaningfully change with the buffer enabled.
With default settings enabling the buffer does not lead to any major changes in memory usage. however if more namespaces are added to the buffer, or if extra process information is collected (through the system module), the numbers can increase

**Review by blakerouse — APPROVED** [review]
Overall this looks really good.

It has great test coverage and it's easy to understand. I feel like I should have found something to comment on, but its really good, nothing I see!

**Comment by ph** [comment]
Going to restart the tests, it been a few months the PR was created.

/test

File: libbeat/monitoring/report/buffer/buffer.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+package buffer
+
+import "sync"
+
+// ringBuffer is a buffer with a fixed number of items that can be tracked.
+//
+// we assume that the size of the buffer is greater than one.
```
Comment (ph) [inline libbeat/monitoring/report/buffer/buffer.go:24]:
```suggestion
// We assume that the size of the buffer is greater than one.
``

File: libbeat/monitoring/report/buffer/reporter.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	}()
+	return r, nil
+}
+
+// Stop will stop the reporter from collecting new information.
+// It will not clear any previously collected data.
+func (r *reporter) Stop() {
+	close(r.done)
+	r.wg.Wait()
+	// Clear entries?
```
Comment (ph) [inline libbeat/monitoring/report/buffer/reporter.go:78]: Remove comment so there is no confusion on the godoc of the function, no need to clear the ring buffer there.

**Review by ph — APPROVED** [review]
Look good to me, adding only a small nit comment.

File: libbeat/api/server.go [hunk 3]
```diff
@@ -72,6 +72,25 @@ func (s *Server) Stop() error {
...
+	defer func() {
+		if r := recover(); r != nil {
+			switch r.(type) {
+			case error:
+				err = r.(error)
+			case string:
+				err = fmt.Errorf(r.(string))
+			default:
+				err = fmt.Errorf("handle attempted to panic with %v", r)
+			}
```
Comment (efd6) [inline libbeat/api/server.go:86]:
```suggestion
			switch r := r.(type) {
			case error:
				err = r
			case string:
				err = errors.New(r)
			default:
				err = fmt.Errorf("handle attempted to panic with %v", r)
			}
```

Needs "errors" added to imports.

File: libbeat/api/server_test.go [hunk 4]
```diff
@@ -184,3 +184,39 @@ func simpleMux() *http.ServeMux {
...
+	h := &testHandler{}
+
+	err = s.AttachHandler("/test", h)
+	require.NoError(t, err)
+
+	r, err := http.Get("http://" + s.l.Addr().String() + "/test")
+	require.NoError(t, err)
+	defer r.Body.Close()
+
+	body, err := ioutil.ReadAll(r.Body)
```
Comment (efd6) [inline libbeat/api/server_test.go:209]: s/ioutil/io/

File: libbeat/monitoring/report/buffer/buffer.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+// the buffer should be thread-safe.
+type ringBuffer struct {
+	entries []interface{}
+	i       int
+	full    bool
+	mu      sync.Mutex
```
Comment (efd6) [inline libbeat/monitoring/report/buffer/buffer.go:30]: Conventionally mutexes are placed above the field they protect.

**Review by efd6 — APPROVED** [review]

**Pull request merged**